### PR TITLE
CDAP-15815: Azure Face Extractor plugin validations.

### PR DESCRIPTION
**Pull request by AntonGensitskiy**
JIRA: https://issues.cask.co/browse/CDAP-15815
WIKI: https://wiki.cask.co/display/CE/Guidelines+on+Plugin+Validation

In scope of this PR:

- integrated with new validation API

### CODE REVIEW

**Comment by AntonGensitskiy** [comment]
@albertshau @CuriousVini could you, please, take a look.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+  public String getFacesSubscriptionKey() {
+    return facesSubscriptionKey;
+  }
+
+  public String getEmotionSubscriptionKey() {
+    return emotionSubscriptionKey;
+  }
+
+  public void validate(FailureCollector failureCollector, Schema inputSchema) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: @Nullable inputSchema

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    if (inputSchema == null) {
+      failureCollector.addFailure("Could not get the input schema to validate.",
+                                  "Please check input schema.");
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: We do not use `Please` in the corrective action. lets keep it consistent
corrective action - `Provide a valid input schema`

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    if (inputSchema.getField(sourceFieldName) == null) {
+      failureCollector.addFailure(
+        String.format("Source field '%s' was not present in the input schema. Schema: '%s'.", sourceFieldName,
+                      inputSchema.toString()), "Provide Source Field that present in the input schema.")
+        .withConfigProperty(SOURCE_FIELD_NAME);
+      return;
+    }
+    if (!inputSchema.getField(sourceFieldName).getSchema().isSimpleOrNullableSimple()) {
+      failureCollector.addFailure(
+        String.format("Source field must be a simple type but was type: '%s'.",
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: error - `Source field '%s' is of unexpected type.`
corrective action - `Ensure field '%s' is of type '%s'`

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                      inputSchema.getField(sourceFieldName).getSchema().getDisplayName()),
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Use displayName on nonnullable schema

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  public void validate(FailureCollector failureCollector, Schema inputSchema) {
+    if (inputSchema == null) {
+      failureCollector.addFailure("Could not get the input schema to validate.",
+                                  "Please check input schema.");
+      return;
+    }
+    if (inputSchema.getField(sourceFieldName) == null) {
+      failureCollector.addFailure(
+        String.format("Source field '%s' was not present in the input schema. Schema: '%s'.", sourceFieldName,
+                      inputSchema.toString()), "Provide Source Field that present in the input schema.")
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: lets not stringify schema in the error message

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+  public String getEmotionSubscriptionKey() {
+    return emotionSubscriptionKey;
+  }
+
+  public void validate(FailureCollector failureCollector, Schema inputSchema) {
+    if (inputSchema == null) {
+      failureCollector.addFailure("Could not get the input schema to validate.",
+                                  "Please check input schema.");
+      return;
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: can just do `throw collector.getOrThrowException()`

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 5 above)
Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: error - `Source field '%s' must be present in the input schema.`
corrective action - null

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        String.format("Source field '%s' was not present in the input schema. Schema: '%s'.", sourceFieldName,
+                      inputSchema.toString()), "Provide Source Field that present in the input schema.")
+        .withConfigProperty(SOURCE_FIELD_NAME);
+      return;
+    }
+    if (!inputSchema.getField(sourceFieldName).getSchema().isSimpleOrNullableSimple()) {
+      failureCollector.addFailure(
+        String.format("Source field must be a simple type but was type: '%s'.",
+                      inputSchema.getField(sourceFieldName).getSchema().getDisplayName()),
+        "Provide Source Field with simple type.")
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: corrective action - `Supported types are : %s`. 
Include list of simple types in above corrective action.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    return emotionSubscriptionKey;
+  }
+
+  public void validate(FailureCollector failureCollector, Schema inputSchema) {
+    if (inputSchema == null) {
+      failureCollector.addFailure("Could not get the input schema to validate.",
+                                  "Please check input schema.");
+      return;
+    }
+    if (inputSchema.getField(sourceFieldName) == null) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Make this a variable and use that every where. `inputSchema.getField(sourceFieldName)`

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        .withConfigProperty(SOURCE_FIELD_NAME);
+    }
+    Schema.Type fieldType = inputSchema.getField(sourceFieldName).getSchema().isNullable()
+      ? inputSchema.getField(sourceFieldName).getSchema().getNonNullable().getType()
+      : inputSchema.getField(sourceFieldName).getSchema().getType();
+
+    if(fieldType != Schema.Type.BYTES) {
+      failureCollector.addFailure(
+        String.format("Input field must be bytes but was: '%s'.",
+                      inputSchema.getField(sourceFieldName).getSchema().getDisplayName()),
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: use `getDisplayName` with non-nullable shcema

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        "Provide Source Field with simple type.")
+        .withConfigProperty(SOURCE_FIELD_NAME);
+    }
+    Schema.Type fieldType = inputSchema.getField(sourceFieldName).getSchema().isNullable()
+      ? inputSchema.getField(sourceFieldName).getSchema().getNonNullable().getType()
+      : inputSchema.getField(sourceFieldName).getSchema().getType();
+
+    if(fieldType != Schema.Type.BYTES) {
+      failureCollector.addFailure(
+        String.format("Input field must be bytes but was: '%s'.",
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: error - `Source field '%s' is of unexpected type '%s'`
corrective action - `Ensure it is of type '%s'`

**Review by CuriousVini — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 1 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 2 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 3 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 4 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 5 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 6 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 7 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 8 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 9 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 10 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

**Comment by AntonGensitskiy** [comment]
@CuriousVini I addressed all your comments, could you, please, take a look.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+  @Macro
+  private String facesSubscriptionKey;
+
+  @Name(EMOTION_SUBSCRIPTION_KEY)
+  @Description("The Azure emotion API subscription key.")
+  @Macro
+  private String emotionSubscriptionKey;
+
+  public AzureFaceExtractorConfig(String sourceFieldName, Boolean continueOnError,
+                String facesSubscriptionKey, String emotionSubscriptionKey) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: nit: indentation

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+      throw failureCollector.getOrThrowException();
+    }
+
+    Schema nonNullableSchema = field.getSchema().isNullable()
+      ? field.getSchema().getNonNullable()
+      : field.getSchema();
+    if (!field.getSchema().isSimpleOrNullableSimple()) {
+      failureCollector.addFailure(
+        String.format("Source field '%s' is of unexpected type.", field.getName()),
+        String.format("Supported type is : %s.", Schema.Type.BYTES))
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: should use displayName

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+        .withConfigProperty(SOURCE_FIELD_NAME);
+    }
+
+    Schema.Type fieldType = nonNullableSchema.getType();
+    if(fieldType != Schema.Type.BYTES) {
+      failureCollector.addFailure(
+        String.format("Source field '%s' is of unexpected type '%s'.",
+                      field.getName(),
+                      nonNullableSchema.getDisplayName()),
+        String.format("Ensure it is of type '%s'", Schema.Type.BYTES))
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: should use displayname

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 11 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 12 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 13 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

**Review by AntonGensitskiy — COMMENTED** [review]
@CuriousVini I addressed all your additional comments, could you, please, take a look.

**Comment by AntonGensitskiy** [comment]
@albertshau @CuriousVini could you, please, take a look.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+      throw failureCollector.getOrThrowException();
+    }
+
+    Schema nonNullableSchema = field.getSchema().isNullable()
+      ? field.getSchema().getNonNullable()
+      : field.getSchema();
+    if (!field.getSchema().isSimpleOrNullableSimple()) {
+      failureCollector.addFailure(
+        String.format("Source field '%s' is of unexpected type.", field.getName()),
+        String.format("Supported type is : '%s'.", Schema.of(Schema.Type.BYTES).getDisplayName()))
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: we can just hard code the corrective action like this: `Ensure it is of type bytes.`

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+        .withConfigProperty(SOURCE_FIELD_NAME);
+    }
+
+    Schema.Type fieldType = nonNullableSchema.getType();
+    if(fieldType != Schema.Type.BYTES) {
+      failureCollector.addFailure(
+        String.format("Source field '%s' is of unexpected type '%s'.",
+                      field.getName(),
+                      nonNullableSchema.getDisplayName()),
+        String.format("Ensure it is of type '%s'", Schema.of(Schema.Type.BYTES).getDisplayName()))
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: we can just hard code the corrective action like this: `Ensure it is of type bytes.`

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+      : field.getSchema();
+    if (!field.getSchema().isSimpleOrNullableSimple()) {
+      failureCollector.addFailure(
+        String.format("Source field '%s' is of unexpected type.", field.getName()),
+        String.format("Supported type is : '%s'.", Schema.of(Schema.Type.BYTES).getDisplayName()))
+        .withConfigProperty(SOURCE_FIELD_NAME);
+    }
+
+    Schema.Type fieldType = nonNullableSchema.getType();
+    if(fieldType != Schema.Type.BYTES) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: also check that its not a decimal logical type

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 14 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 15 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 16 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: Done.

**Comment by AntonGensitskiy** [comment]
@CuriousVini I addressed all your additional comments, could you, please, take a look.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+        .withConfigProperty(SOURCE_FIELD_NAME);
+      throw failureCollector.getOrThrowException();
+    }
+
+    Schema nonNullableSchema = field.getSchema().isNullable()
+      ? field.getSchema().getNonNullable()
+      : field.getSchema();
+    if (!field.getSchema().isSimpleOrNullableSimple()) {
+      failureCollector.addFailure(
+        String.format("Source field '%s' is of unexpected type.", field.getName()),
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: `Source field '%s' is of unexpected type '%s'.`

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+      failureCollector.addFailure(
+        String.format("Source field '%s' must be present in the input schema.", sourceFieldName), null)
+        .withConfigProperty(SOURCE_FIELD_NAME);
+      throw failureCollector.getOrThrowException();
+    }
+
+    Schema nonNullableSchema = field.getSchema().isNullable()
+      ? field.getSchema().getNonNullable()
+      : field.getSchema();
+    if (!field.getSchema().isSimpleOrNullableSimple()) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: we dont need this check right?

**Review by CuriousVini — APPROVED** [review]
2 comments. Fix and squash.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 17 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: After all changes in validation , this check is redundant.
Removed.

File: src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java (hunk 18 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/azure/face/extractor/AzureFaceExtractorConfig.java]: After all changes in validation , this check is redundant.
Removed.

**Comment by AntonGensitskiy** [comment]
Commits are squashed.

**Review by CuriousVini — APPROVED** [review]

**Pull request merged**